File: stack/leetcode_Min_Stack.java
/**
 * kleetcode_Min_Stack.
 * leetcode
 *
 * Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.
 *
 * push(x) -- Push element x onto stack.
 * pop() -- Removes the element on top of the stack.
 * top() -- Get the top element.
 * getMin() -- Retrieve the minimum element in the stack.
 */

import java.util.ArrayList;
import java.util.List;


public class leetcode_Min_Stack {

  private final List<Integer> stack;
  private final List<Integer> min;

  private int length;

  public leetcode_Min_Stack() {
    this.stack = new ArrayList<Integer>();
    this.min = new ArrayList<Integer>();
  }

  public void push(int x) {
    this.stack.add(x);
    if (this.min.isEmpty() || this.min.get(this.min.size() - 1) >= x) {
      this.min.add(x);
    }
  }

  public void pop() {
    int elem = this.stack.remove(stack.size() - 1);
    int length = this.min.size();
    if (elem == this.min.get(length - 1)) {
      this.min.remove(length - 1);
    }
  }

  public int top() { return this.stack.get(this.stack.size()-1); }

  public int getMin() { return this.min.get(this.min.size()-1); }
}


/** Python solution.

class MinStack(object):

    def __init__(self):
        """"""
        self._s = []
        self._min = []

    def push(self, x):
        """"""
        self._s.append(x)

        if not self._min or self._min[-1] >= x:
            self._min.append(x)

    def pop(self):
        """"""
        x = self._s.pop()
        if self._min[-1] == x:
            self._min.pop()

    def top(self):
        """"""
        return self._s[-1]

    def getMin(self):
        return self._min[-1]
*/

File: stack/leetcode_Min_Stack.py
"""Min stack
leetcode

Design a stack that supports push, pop, top, and retrieving the minimum element
 in constant time.

push(x) -- Push element x onto stack.
pop() -- Removes the element on top of the stack.
top() -- Get the top element.
getMin() -- Retrieve the minimum element in the stack.
"""


class MinStack(object):

    def __init__(self):
        """
        initialize your data structure here.
        """
        self._stack = []
        self._mins = []

    def push(self, x):
        """
        :type x: int
        :rtype: nothing
        """
        if self._stack:
            prev_min, count = self._mins[-1]
            if x < prev_min:
                self._mins.append([x, 1])
            else:
                self._mins[-1][1] += 1

        else:
            self._mins.append([x, 1])
        self._stack.append(x)

    def pop(self):
        """
        :rtype: nothing
        """
        if self._mins[-1][1] > 1:
            self._mins[-1][1] -= 1
        elif self._mins[-1][1] == 1:
            self._mins.pop()
        else:
            raise ValueError
        return self._stack.pop()

    def top(self):
        """
        :rtype: int
        """
        return self._stack[-1]

    def getMin(self):
        """
        :rtype: int
        """
        return self._mins[-1][0]


def main():
    mstack = MinStack()
    mstack.push(-2)
    mstack.push(0)
    mstack.push(-1)

    assert mstack.getMin() == -2

    mstack.push(-3)
    assert mstack.getMin() == -3
    mstack.push(-10)
    assert mstack.getMin() == -10

    assert mstack.pop() == -10
    assert mstack.getMin() == -3

    assert mstack.pop() == -3
    assert mstack.getMin() == -2

    assert mstack.pop() == -1
    assert mstack.getMin() == -2

    assert mstack.pop() == 0
    assert mstack.getMin() == -2

    assert mstack.pop() == -2


if __name__ == '__main__':
    main()
